File: ts/src/parkingBoy/mostAvailableParkingBoyService.ts
/** @format */

import {ParkingBoyService} from './parkBoyService'
import {ParkingLot} from '../parkingLot/parkingLot'

export class MostAvailableParkingBoyService implements ParkingBoyService {
    public findOneValidParkingLot(parkingLots: ParkingLot[]) {
        return (
            parkingLots
                .filter(({hasAvailableSpaces}) => hasAvailableSpaces)
                .sort(({availableSpaces}) => -availableSpaces)[0] || null
        )
    }
}

File: kt/src/main/kotlin/com/github/scglwsj/parking/boy/MostAvailableParkingBoyService.kt
package com.github.scglwsj.parking.boy

import com.github.scglwsj.parking.lot.ParkingLot

class MostAvailableParkingBoyService : ParkingBoyService {
    override fun findOneValidParkingLot(parkingLots: List<ParkingLot>): ParkingLot? =
        parkingLots.sortedByDescending { it.availableSpaces }.firstOrNull { it.hasAvailableSpaces }
}
